frontend-react/docs/test-conventions.md

```diff
@@ -0,0 +1,492 @@
+        });
+
+        test("renders the provided message", () => {
+            expect(screen.getByText("Hello!")).toBeVisible();
+        });
+        
+        test("does not render the fallback message", () => {
+            expect(screen.queryByText("This is a fallback message.")).toBeNull();
+        });    
+    });
+});
+```
+
+> Why?
+>
+> It's redundant to test third-party library logic, even implicitly, because that logic is entirely out of our control and really should already be tested by the maintainers of said library.
+
+---
+
+## Test Practices
+
+### Use `before*` and `after*` hooks for setup/teardown and `test`/`it` for expectations
+
+```tsx
+// bad
+describe("LoginForm", () => {
+    test("shows the welcome message", async () => {
+        jest.spyOn(someModule, "getWelcomeMessage").mockReturnValue("Welcome!");
+    
+        render(<LoginFor />);
+    
+        await userEvent.type(emailInputNode, "abc@123.com");
+        await userEvent.type(passwordInputNode, "password");
+
+        expect(screen.getByText("Welcome!")).toBeVisible();
+        
+        jest.restoreAllMocks();
+    });
+});
+
+// good
+describe("LoginForm", () => {
+    let emailInputNode;
+    let passwordInputNode;
+
+    beforeEach(async () => {
+        jest.spyOn(someModule, "getWelcomeMessage").mockReturnValue("Welcome!");
+        jest.spyOn(someModule, "getDeniedMessage").mockReturnValue("Denied!");
+
+        render(<LoginForm />);
+    
+        await userEvent.type(emailInputNode, "abc@123.com");
+        await userEvent.type(passwordInputNode, "password");   
+    });
+    
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+    
+    test("shows the welcome message", () => {
+        expect(screen.getByText("Welcome!")).toBeVisible();
+    });
+});
+```
+
+> Why?
+>
+> `before` and `after` are specifically meant for setup and teardown, respectively, and they should be used as such to keep each piece of any given test focused -- most notably, it encourages running more granular expectations under the same conditions.  This also allows us to add more granular expectations when new features (or regressions) need to be handled.
+
+#### Use `describe`s to separate conditions
+
+```tsx
+// bad
+describe("LoginForm", () => {
+    test("clears the form when the user clicks 'Continue'", () => {
+        ...
+    });
+    
+    test("clears the form when the user clicks 'Cancel'", () => {
+        ...
+    });
+});
+
+// good
+describe("LoginForm", () => {
+    describe("when the user clicks 'Continue'", () => {
+        test("clears the form", () => {
+            ...
+        });
+    });
+
+    describe("when user clicks 'Cancel'", () => {
+        test("clears the form", () => {
+            ...
+        });
+    });
+});
+```
+
+> Why?
+>
+> By organizing our tests into sets of conditions, it becomes much easier to add in more expectations in the future to account for new features and regressions.
```

Comment:
It might be important to note that before* and after* functions are scoped to the parent describe (at least I think with some experience I had with the msw update).

Reply:
Good point!  Updated [here](https://github.com/CDCgov/prime-reportstream/pull/8430/files#diff-42fae80a7a121e5a9d806d505777c76a86e871122ef5bb86be2646e4ae4e5cf6R409).  Let me know if you think even more clarity is needed here!